File: src/core/Notes/Import.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using core.Adapters.CSV;
using core.Shared;
using MediatR;

namespace core.Notes
{
    public class Import
    {
        public class Command : RequestWithUserId<CommandResponse>
        {
            public Command(string content)
            {
                this.Content = content;
            }

            public string Content { get; }
        }

        public class Handler : MediatR.IRequestHandler<Command, CommandResponse>
        {
            private IMediator _mediator;
            private ICSVParser _parser;

            public Handler(IMediator mediator, ICSVParser parser)
            {
                _mediator = mediator;
                _parser = parser;
            }

            public async Task<CommandResponse> Handle(Command request, CancellationToken cancellationToken)
            {
                var records = _parser.Parse<NoteRecord>(request.Content);

                foreach(var r in records)
                {
                    var c = new core.Notes.Add.Command{
                        Note = r.note,
                        Ticker = r.ticker,
                        Created = r.created,
                    };

                    c.WithUserId(request.UserId);

                    var ar = await _mediator.Send(c);

                    if (ar.Error != null)
                    {
                        return ar;
                    }
                }

                return CommandResponse.Success();
            }

            private class NoteRecord
            {
                public DateTimeOffset created { get; set; }
                public string ticker { get; set; }
                public string note { get; set; }
            }
        }
    }
}

File: src/core.fs/Options/Import.fs
namespace core.fs.Options

open System
open core.fs
open core.fs.Accounts
open core.fs.Adapters.Brokerage
open core.fs.Adapters.CSV

module Import =
    open core.Shared

    type OptionRecord = {
        ticker:string
        strike:decimal
        optiontype:string
        expiration:Nullable<DateTimeOffset>
        amount:int
        premium:decimal
        filled:DateTimeOffset
        ``type``:string
    }
          
    type Command(content:string,userId:UserId) =
        member _.UserId = userId
        member _.Content = content

    type Handler(csvParser:ICSVParser, optionHandler:core.fs.Options.OptionsHandler) =
        
        member _.Handle (request:Command) token = task {
            let parseResponse = csvParser.Parse<OptionRecord>(request.Content)

            match parseResponse with
            | Error err -> return Error err
            | Ok _ -> return "Not implemented" |> ServiceError |> Error
        }
            
